File: Minimum number of jumps/sol.py
class Solution:
    def minJumps(self, arr, n):
        if arr[0] == 0:
            return -1
        if n == 1:
            return 0
        
        maxReach = arr[0]
        steps = arr[0]
        jumps = 1
        
        for i in range(1, n):
            maxReach = max(maxReach, i + arr[i])
            steps -= 1
            
            if i == n - 1:
                return jumps
            
            if steps == 0:
                jumps += 1
                
                if i >= maxReach:
                    return -1
                
                steps = maxReach - i
        
        return -1

File: Minimum number of jumps/sol.cpp
class Solution{
  public:
    int minJumps(int arr[], int n){
        if(arr[0] == 0){
            return -1;
        }
        if(n==1){
            return 0;
        }
        
        int maxReach = arr[0];
        int steps = arr[0];
        int jumps = 1;
        
        for(int i=1; i<n; i++){
            maxReach = max(maxReach, i + arr[i]);
            steps--;
            if(i==n-1){
                return jumps;
            }
            
            if(steps==0){
                jumps++;
                
                if(i>=maxReach){
                    return -1;
                }
                
                steps = maxReach - i;
            }
            
        }
        
        return -1;
        }
    
};
